File: workspace/2023/2023-02-10/run-length-encoding.ruby/run_length_encoding_test.rb
require 'test/unit'
require 'run_length_encoding'

class TestCase < Test::Unit::TestCase

    TEST_DATA_LISTS = [
        # [xs, runs, encoded_runs]
        [[],             [],                                   []],
        ['a'.chars,      ['a'.chars],                          [['a', 1]]],
        ['aa'.chars,     ['aa'.chars],                         [['a', 2]]],
        ['aaa'.chars,    ['aaa'.chars],                        [['a', 3]]],
        ['ab'.chars,     ['a'.chars, 'b'.chars],               [['a', 1], ['b', 1]]],
        ['abb'.chars,    ['a'.chars, 'bb'.chars],              [['a', 1], ['b', 2]]],
        ['aabccc'.chars, ['aa'.chars, 'b'.chars, 'ccc'.chars], [['a', 2], ['b', 1], ['c', 3]]],
    ]

    TEST_DATA_LISTS.each_with_index do |(xs, runs, encoded_runs), i|
        define_method("test_encode_decode_#{i}") do
            assert_equal(xs, decode(encode(xs)))
        end
    end

    TEST_DATA_LISTS.each_with_index do |(xs, runs, encoded_runs), i|
        define_method("test_encode_#{i}") do
            assert_equal(encoded_runs, encode(xs))
        end
    end

    TEST_DATA_LISTS.each_with_index do |(xs, runs, encoded_runs), i|
        define_method("test_decode_#{i}") do
            assert_equal(xs, decode(encoded_runs))
        end
    end

end

File: workspace/2023/2023-02-01/run-length-encoding.python/run_length_encoding_test.py
import unittest
from run_length_encoding import encode, decode
from parameterized_testing import parameterized


class TestCase(unittest.TestCase):

    TEST_DATA_LISTS = [
        # (xs, runs, encoded_runs)
        ([],                [],                                     []),
        (list("a"),         [list("a")],                            [("a", 1)]),
        (list("aa"),        [list("aa")],                           [("a", 2)]),
        (list("aaa"),       [list("aaa")],                          [("a", 3)]),
        (list("ab"),        [list("a"), list("b")],                 [("a", 1), ("b", 1)]),
        (list("abb"),       [list("a"), list("bb")],                [("a", 1), ("b", 2)]),
        (list("aabccc"),    [list("aa"), list("b"), list("ccc")],   [("a", 2), ("b", 1), ("c", 3)]),
    ]

    @parameterized(TEST_DATA_LISTS)
    def test_encode(self, xs, runs, encoded_runs):
        self.assertEqual(encoded_runs, encode(xs))

    @parameterized(TEST_DATA_LISTS)
    def test_decode(self, xs, runs, encoded_runs):
        self.assertEqual(xs, decode(encoded_runs))


if __name__ == "__main__":
    unittest.main()

